server/api/controllers/volunteer-application-controller.ts

```diff
@@ -0,0 +1,147 @@
+import nodemailer from "nodemailer";
+import validator from "validator";
+import { Request, Response } from "express";
+
+import { Logger } from "../../util/logger";
+
+export const emailDomain = "@mathsoc.uwaterloo.ca";
+
+export class VolunteerApplicationController {
+  static logger = new Logger();
+
+  static handleRequest(req: Request, res: Response) {
+    if (!req.body || !req.headers.referer) {
+      res.status(400).end();
+      return false;
+    }
+
+    if (
+      !this.validateText(req.body["first-name"], 100) ||
+      !(req.body["preferred-name"].length > 100) ||
+      !this.validateText(req.body["last-name"], 100) ||
+      !this.validateText(req.body.email, 100) ||
+      !validator.isEmail(req.body.email) ||
+      !this.validateText(req.body.interest, 6000) ||
+      !this.validateText(req.body.qualifications, 6000) ||
+      !(req.body["more-info"].length > 6000)
+    ) {
+      res.status(400).end();
+      this.logger.info(
+        "Volunteer application email attempted with invalid parameters; not sent"
+      );
+      return false;
+    }
+
+    const queryParams = req.headers.referer.split("?")[1];
+    const formBody = {
+      firstName: validator.escape(req.body["first-name"]),
+      preferredName: validator.escape(req.body["preferred-name"]),
+      lastName: validator.escape(req.body["last-name"]),
+      program: validator.escape(req.body.program),
+      term: validator.escape(req.body.term),
+      coop: validator.escape(req.body.coop),
+      address: validator.escape(req.body.email),
+      interest: validator.escape(req.body.interest),
+      qualifications: validator.escape(req.body.qualifications),
+      moreInfo: validator.escape(req.body["more-info"]),
+      role: VolunteerApplicationController.getRoleFromQueryParams(queryParams),

```

Comment:
should `role` and `execAddress` also be escaped?

Reply:
it won't have an affect at the moment since it only escapes certain characters [docs](https://github.com/validatorjs/validator.js#sanitizers) but added in case that changes 👍